### ISSUE TITLE
10 million samples limit

### ISSUE THREAD

**Issue by xtemp09** [tenacityteam/tenacity-legacy#485]
### Guidelines

- [X] I have read the guidelines.

### Is your feature request related to a problem? Please describe.

I'm always frustrated with [10 million samples limit](https://github.com/tenacityteam/tenacity/blob/master/src/FreqWindow.cpp#L670).

### Describe the solution you'd like

My frustration can be removed with removal of this limit.

### Describe alternatives you've considered

The alternative is to download, fix the code and compile it on my own.

### This issue is not a duplicate

- [X] I have confirmed this issue isn't a duplicate.

**Comment by petersampsonaudacity** [comment]
I see no such limit
![image](https://user-images.githubusercontent.com/50205984/129761557-e18de81b-41e2-44c7-ba91-8cc0f0107c35.png)

Sure the Time Toolbar resets to 000,000,000 after 999,999,999 samples - but there is no limit

**Comment by Paul-Licameli** [comment]
Referring to Analyze > Plot Spectrum I think

**Comment by Be-ing** [comment]
> The alternative is to download, fix the code and compile it on my own.

Pull requests welcome.

**Comment by Be-ing** [comment]
I presume that hardcoded limit is there for a reason. I am not familiar with that code and there is no comment explaining why that limit is there so I am concerned what would happen if the limit was removed. @nyanpasu64 do you know why that limit is there?

**Comment by nyanpasu64** [comment]
- https://github.com/tenacityteam/tenacity/blob/db09547873ec4153860e0c1b37af6d4d26070a94/src/FreqWindow.cpp#L692
- https://github.com/tenacityteam/tenacity/blob/db09547873ec4153860e0c1b37af6d4d26070a94/src/SampleFormat.h#L174
- https://github.com/tenacityteam/tenacity/blob/db09547873ec4153860e0c1b37af6d4d26070a94/libraries/lib-utility/MemoryX.h#L31

My guess is that the sample limit is placed to avoid allocating too much memory for the frequency analysis tool which could cause the computer to OOM or Audacity to crash.

You could remove the limit. But in my (reasonably experienced but not authoritative) opinion, I think the better approach to fixing this bug is to use a smaller FFT buffer, either the size of a single FFT and replenished on every FFT, or longer and replenished with data from the main document whenever it comes close to running out. This will make the memory required constant regardless of how much audio you're analyzing.

This would probably require rewriting [`SpectrumAnalyst::Calculate(...)`](https://github.com/tenacityteam/tenacity/blob/db09547873ec4153860e0c1b37af6d4d26070a94/src/SpectrumAnalyst.cpp#L93) to not take `const float *data` but something else, perhaps a callback (interface or std::function or tl::function_ref) which returns a pointer to `windowSize` floats. This could be a 1-window-long buffer rewritten every call, or a pointer within a larger buffer internally that only gets rewritten once depleted (this means less writes total because adjacent buffers overlap by 50% due to `start += half`). Note that this function is also used by [`SelectHandle::StartSnappingFreqSelection(...)`](https://github.com/tenacityteam/tenacity/blob/db09547873ec4153860e0c1b37af6d4d26070a94/src/tracks/ui/SelectHandle.cpp#L1385) (I don't know what uses this code), which also has a limited size (10485760), so more code to rewrite, lovely...

However I don't have the motivation right now to fix the bug. You could remove the limit as a stopgap I suppose.

**Comment by xtemp09** [comment]
Yes, the limit shows up in Plot Spectrum. I usually increased the limit, recently, I have reinstalled Windows and changed toolchain, so, now I don't compile Audacity on my own.

![limit](https://user-images.githubusercontent.com/9303235/129839146-3291ff23-b9c7-4464-9a1b-f68ba7019d07.png)

The good solution is to rewrite the code, perhaps, as nyanpasu64 mentioned.

**Comment by xtemp09** [comment]
When I calculate hashsum, I load a chunk of data, process it, then load another chunk ― pretty much 99% of hashsum calculators do. Audacity has overcome my expectations and load only the first 237 seconds, leaving the other ones.

**Comment by nyanpasu64** [comment]
Out of curiosity, why are you taking the spectrum of hour-long audio signals? (Though 237 seconds does seem short.)

And is 65536 samples long enough of a window?

**Comment by xtemp09** [comment]
It was just an example, I usually process short audio files (4-5 minutes). I used to double the limit.

**Comment by xtemp09** [comment]
Here is another example:

![limit2](https://user-images.githubusercontent.com/9303235/129841257-34e6b925-ce82-4225-bd64-3d1a5cb8e08c.png)

**Comment by Be-ing** [comment]
> When I calculate hashsum

Why are you hashing audio signals? I'm not sure Tenacity nor Audacity are the right tool for the job. Have you considered [Chromaprint](https://github.com/acoustid/chromaprint)?

**Comment by n0toose** [comment]
> > When I calculate hashsum
> 
> Why are you hashing audio signals? I'm not sure Tenacity nor Audacity are the right tool for the job.

I've used Audacity/Tenacity to bitcrush or even analyze binaries. Leaving tool recommendations is nice, but there's a chance that people still use the audio editor for good reason. It's a general purpose tool, they probably won't be the last to recommend it.

**Comment by nyanpasu64** [comment]
I'm more confused why @xtemp09 is using the frequency analyzer to calculate a hashsum of a file. Unless you're using it for something other than a checksum, like gathering statistics on a file?

**Comment by n0toose** [comment]
Apart from finding out for the sake of fulfilling our curiosity, it may be best to not let the use case in particular here get in the way of making an accommodation and getting rid of an arbitrarily caused problem, considering that @xtemp09 will most likely not be the last person to ever need this.

**Comment by xtemp09** [comment]
@nyanpasu64, thank you for the witty and amusing comment. For those ones who did not understand ― it was an example of processing data. Load a chunk of data, process it, load another one, process it. The limitation is stupid.

I'll write it down in bold:
### Having this limit sucks. Period.

And off topic: `Audacity | s/Aud/Ten/g` can be used for other things too. I converted signals (non-audio) and used denoiser in Audacity, then converted the output files back. Nowadays, I use Matlab to denoise.

Another off topic: if you convert losslessly, files are different only in metadata. So, if there is no metadata, md5sums coincide. That's how I found a bug in FFMPEG. FFMPEG added additional chunk of silence in the beginning of MP4 file. I would provide you a link, but https://trac.ffmpeg.org seems to be down.
